File: Leetcode/130.cpp
namespace DSU {
    vector<int> father;
    vector<int> Size;
    void init(int n) {
        vector<int>().swap(father);
        vector<int>().swap(Size);
        father.resize(n + 5);
        iota(father.begin(), father.end(), 0);
        Size.resize(n + 5, 1);
    }
    int find(int x) {
        if (father[x] != x) father[x] = find(father[x]);
        return father[x];
    }
    void merge(int x, int y) {
        x = find(x);
        y = find(y);
        if (x == y) return;
        if (Size[x] > Size[y]) swap(x, y);
        Size[y] += Size[x];
        father[x] = y;
    }
}

class Solution {
public:
    void solve(vector<vector<char>>& board) {
        if (board.empty()) return;
        constexpr int STEPS = 4;
        constexpr int dx[] = {-1, 1, 0, 0};
        constexpr int dy[] = {0, 0, -1, 1};
        int n = (int)board.size();
        int m = (int)board.front().size();
        DSU::init(n * m);
        int new_i, new_j;
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                if (board.at(i).at(j) == 'O') {
                    for (int step = 0; step < STEPS; ++step) {
                        new_i = i + dx[step];
                        new_j = j + dy[step];
                        if (new_i >= 0 && new_i < n && new_j >= 0 && new_j < m && board.at(new_i).at(new_j) == 'O') {
                            DSU::merge(new_i * m + new_j, i * m + j);
                        }
                    }
                }
            }
        }
        unordered_set<int> groupsTouchEdge;
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                if ((i == 0 || j == 0 || i == n - 1 || j == m - 1) && board.at(i).at(j) == 'O') {
                    groupsTouchEdge.insert(DSU::find(i * m + j));
                }
            }
        }
        // flipping
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                if (board.at(i).at(j) == 'O' && groupsTouchEdge.find(DSU::find(i * m + j)) == groupsTouchEdge.end()) {
                    board.at(i).at(j) = 'X';
                }
            }
        }
    }
};

File: Leetcode/130/130.go
const (
	STEPS = 4
)
var (
	dx = [...]int{-1,1,0,0}
	dy = [...]int{0,0,-1,1}
)

// DSU
var (
	father []int
	Size []int
)

func initDsu(n int)  {
	father = make([]int, n+5)
	for i := 0; i < len(father); i++ {
		father[i] = i
	}
	Size = make([]int, n+5)
	for i := 0; i < len(Size); i++ {
		Size[i] = 1
	}
}
func find(x int) int {
	if x != father[x] {
		father[x] = find(father[x])
	}
	return father[x]
}
func merge(x, y int) {
	x = find(x)
	y = find(y)
	if x == y {
		return
	}
	if Size[x] > Size[y] {
		x, y = y, x
	}
	Size[y] += Size[x]
	father[x] = y
}

func solve(board [][]byte)  {
	n := len(board)
	m := len(board[0])
	initDsu(n*m)
	var newI, newJ int
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if board[i][j] == 'O' {
				for step := 0; step < STEPS; step++ {
					newI = i + dx[step]
					newJ = j + dy[step]
					if newI >= 0 && newI < n && newJ >= 0 && newJ < m && board[newI][newJ] == 'O' {
						merge(i*m+j, newI*m+newJ)
					}
				}
			}
		}
	}
	groups := make(map[int][]int)
	var groupId int
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if board[i][j] == 'O' {
				groupId = find(i*m+j)
				groups[groupId] = append(groups[groupId], i*m+j)
			}
		}
	}
	var haveAns bool
	var r, c int
	for _, nodeList := range groups {
		haveAns = true
		for i := 0; i < len(nodeList); i++ {
			r = nodeList[i]/m
			c = nodeList[i]%m
			if r == 0 || r == n-1 || c == 0 || c == m-1 {
				haveAns = false
				break
			}
		}
		if haveAns {
			for i := 0; i < len(nodeList); i++ {
				r = nodeList[i]/m
				c = nodeList[i]%m
				board[r][c] = 'X'
			}
		}
	}
}